planner/core/expression_rewriter.go

```diff
@@ -943,10 +943,38 @@ func (er *expressionRewriter) isNullToExpression(v *ast.IsNullExpr) {
 }
 
 func (er *expressionRewriter) positionToScalarFunc(v *ast.PositionExpr) {
-	if v.N > 0 && v.N <= er.schema.Len() {
-		er.ctxStack = append(er.ctxStack, er.schema.Columns[v.N-1])
-	} else {
-		er.err = ErrUnknownColumn.GenWithStackByArgs(strconv.Itoa(v.N), clauseMsg[er.b.curClause])
+	stkLen := len(er.ctxStack)
+	if val, ok := v.N.(*driver.ValueExpr); ok {
+		pos, isPos := val.GetValue().(int64)
+		if isPos && pos > 0 && pos <= int64(er.schema.Len()) {
+			er.ctxStack = er.ctxStack[:stkLen-1]
+			er.ctxStack = append(er.ctxStack, er.schema.Columns[pos-1])
+			return
+		}
+		er.err = ErrUnknownColumn.GenWithStackByArgs(strconv.Itoa(int(pos)), clauseMsg[er.b.curClause])
+	}
+	if val, ok := v.N.(*driver.ParamMarkerExpr); ok {
+		if len(er.ctx.GetSessionVars().PreparedParams) == 0 {
+			return
+		}
+		if param, err := expression.GetParamExpression(er.ctx, val, er.useCache()); err == nil {
+			if con, ok := param.(*expression.Constant); ok {
+				str, isNull, err := con.EvalString(er.ctx, chunk.Row{})
```

Comment:
Can we use `Constant::EvalInt` here?

Reply:
We can't. This is because``Constant::EvalInt`` does not return an error when trying to convert a non numeric string (e..g., ``"##"``). That is, for example, it returns ``0`` without an error when converting ``"##"``.